apps/build-onchain-apps/contracts/src/SignatureMintERC721.sol

```diff
@@ -0,0 +1,143 @@
+        bytes memory jsonMetadata = abi.encodePacked(
+            "{",
+            NFTUtilities.getEncodedStringPair("description", metadata.description, false),
+            NFTUtilities.getEncodedStringPair("symbol", metadata.symbol, false),
+            NFTUtilities.getEncodedStringPair("image", metadata.image, false),
+            NFTUtilities.getEncodedStringPair("animation_url", metadata.animation_url, true),
+            "}"
+        );
+        return string(jsonMetadata);
+    }
+
+    function tokenURI(uint256 tokenId) public view override(ERC721) returns (string memory) {
+        require(tokenId <= currentId, "Token doesnt exist");
+        return getOpenEditionUri(tokenId);
+    }
+
+    /**
+     * @dev Function to return the token metadata, this is stored on contract versus IPFS.  This is a common practice
+     *      for NFT collections which all have the same metadata (a.k.a OpenEditions).
+     */
+    function getOpenEditionUri(uint256 token) public view returns (string memory) {
+        // Generates a name with edition .. e.g. NFT => NFT #1029
+        string memory nameWithTokenId = NFTUtilities.getEncodedStringPair(
+            "name", string(abi.encodePacked(metadata.name, " #", Strings.toString(token))), false
+        );
+
+        bytes memory jsonMetadata = abi.encodePacked(
+            "{",
+            nameWithTokenId,
+            NFTUtilities.getEncodedStringPair("description", metadata.description, false),
+            NFTUtilities.getEncodedStringPair("symbol", metadata.symbol, false),
+            NFTUtilities.getEncodedStringPair("image", metadata.image, false),
+            NFTUtilities.getEncodedStringPair("animation_url", metadata.animation_url, true),
+            "}"
+        );
+        return string(jsonMetadata);
+    }
+
+    /**
+     * @dev Helper method to get the hash(bytes) needed to be signed for an account to get a free mint.
+     * @param  account that will receive the NFT
+     */
+    function getBytesToSign(address account) external pure returns (bytes32) {
+        bytes memory messageEncodedPacked = abi.encodePacked(getStringToSign(account));
+        return MessageHashUtils.toEthSignedMessageHash(messageEncodedPacked);
+    }
+
+    /**
+     * @dev Helper method to get the string needed to be signed by a web3 wallet.
+     * @param  account that will receive the NFT
+     */
+    function getStringToSign(address account) public pure returns (string memory) {
+        return string(abi.encodePacked("Free Mint: ", Strings.toHexString(uint256(uint160(account)), 20)));
+    }
+
+    /**
+     * @dev Verifies a message was signed by the freeMintSigner wallet.
+     * @param  account that will receive the NFT
+     * @param  signature authorizing the action
+     */
+    function verifySignature(address account, bytes memory signature) public view returns (bool) {
+        bytes32 ethSignedMessageHash = this.getBytesToSign(account);
```

Comment:
ethSigned message usually refers to something ERC 191 compliant, which this is not?

Reply:
but would probably be better if it followed the ERC 191 or ERC 712

Reply:
This uses the same naming that OpenZepplin uses : https://docs.openzeppelin.com/contracts/2.x/utilities... AFAIK that is following ERC191 as we are using MessageHashUtils.toEthSignedMessageHash(messageEncodedPacked); in  getBytesToSign

Reply:
oops sorry missed that, is right!